ludwig/models/llm.py

```diff
@@ -33,16 +34,40 @@ def __init__(
         self,
         config_obj: LLMModelConfig,
         random_seed=None,
+        device=None,
         **_kwargs,
     ):
         super().__init__(random_seed=random_seed)
 
         self.config_obj = config_obj
         self._random_seed = random_seed
 
+        self.model_name = self.config_obj.model_name
+
+        logger.info("Loading large language model...")
         self.model = AutoModelForCausalLM.from_pretrained(self.config_obj.model_name)
-        self.generation_config = GenerationConfig(**self.config_obj.generation_config.to_dict())
-        self.max_new_tokens = self.config_obj.generation_config.max_new_tokens
+        self.curr_device = torch.device("cpu")  # model initially loaded onto cpu
+        logger.info("Done.")
+
+        if device is None:
+            device = get_torch_device()
+        self.to_device(device)
+
+        self.max_new_tokens = self.config_obj.generation.max_new_tokens
+
+        # Determines the maximum length of the context (input + output tokens)
+        if hasattr(self.model.config, "max_sequence_length"):
+            self.context_len = self.model.config.max_sequence_length
+        elif hasattr(self.model.config, "max_position_embeddings"):
+            self.context_len = self.model.config.max_position_embeddings
+        else:
+            self.context_len = 2048
+        self.max_input_length = self.context_len - self.max_new_tokens - 8
+
+        # Used only for its metadata about the vocabulary
+        self.tokenizer = AutoTokenizer.from_pretrained(self.config_obj.model_name, use_fast=False)
```

Comment:
Just as we discussed, I think we can set `use_fast` to True if we set the protobuf dependency to be pinned to `3.20.3` with `tokenizers` set to 0.13.3. But also ok to leave it like this for now with a TODO to come back to making this `use_fast` by default.

Reply:
Also damn, do we have to initialize the tokenizer just to grab the `bos_token_id`? That kinda sucks and is super unfortunate

Reply:
Yeah that's right– because its just for the `bos_token_id`, maybe I'll just leave it as-is since it will load much faster. Here's the Linear ticket to improve this downstream: https://linear.app/predibase/issue/MLX-1082/add-tokenizer-cache-to-ensure-that-fast-llamatokenizer-is-only-loaded